spec/furniture/marketplace/cart_products_controller_request_spec.rb

```diff
@@ -5,36 +5,117 @@
   let(:space) { marketplace.space }
   let(:room) { marketplace.room }
   let(:member) { create(:membership, space: space).member }
+  let(:document_root_element) { Nokogiri::HTML::Document.parse(response.body) }
+  let(:request_as) { :html }
 
   describe "#create" do
+    subject(:perform_request) do
+      post path, as: request_as, params: {cart_product: {product_id: product.id, quantity: 1}}
+      response
+    end
+
+    let(:product) { create(:marketplace_product, marketplace: marketplace) }
+    let(:cart) { create(:marketplace_cart, marketplace: marketplace) }
+    let(:path) { polymorphic_path(cart.location(child: :cart_products)) }
+
     it "Add a Product to the Cart" do
-      product = create(:marketplace_product, marketplace: marketplace)
-      cart = create(:marketplace_cart, marketplace: marketplace)
+      perform_request
 
-      post polymorphic_path(cart.location(child: :cart_products), params: {cart_product: {product_id: product.id}})
       expect(response).to redirect_to([space, room])
     end
+
+    context "when a turbo stream" do
+      let(:request_as) { :turbo_stream }
+
+      it "Replaces the cart product, cart footer and cart total" do
+        perform_request && cart.reload
+
+        assert_select("turbo-stream[action='replace'][target='cart-product-#{product.id}']") do
+          assert_select("*[data-cart-product-quantity]", text: "1")
+        end
+
+        assert_select("turbo-stream[action='replace'][target='cart-footer-#{cart.id}']")
+        assert_select("turbo-stream[action='replace'][target='cart-total-#{cart.id}'] *[data-cart-total]", text: "Total: #{controller.helpers.humanized_money_with_symbol(cart.price_total)}")
```

Comment:
Maybe the `have_rendered_turbo_stream` matcher could take a block, and then you could pass in some code to match the content?

Reply:
Yea,  I've been looking at the shoulda-matchers api as a potential inspiration for a Turbo/ViewComponent/Rails-Dom-Testing oriented matcher set; but I hadn't really fiddled with it enough to have a concrete proposal. Maybe I'll putz about at it later now that we're getting a wider set of request and component specs.